File: 05-hashmap/07-happy-number/Solution.java
import java.util.HashSet;
import java.util.Set;

class Solution {
    private int nextVal(int x) {
        int total = 0;
        while (x > 0) {
            int digit = x % 10;
            total += digit * digit;
            x /= 10;
        }
        return total;
    }

    public boolean isHappy(int n) {
        Set<Integer> seen = new HashSet<>();
        while (n != 1 && !seen.contains(n)) {
            seen.add(n);
            n = nextVal(n);
        }
        return n == 1;
    }
}

File: 05-hashmap/07-happy-number/Solution.py
class Solution:
    def isHappy(self, n: int) -> bool:
        def next_val(x: int) -> int:
            total = 0
            while x > 0:
                digit = x % 10
                total += digit * digit
                x //= 10
            return total

        seen = set()
        while n != 1 and n not in seen:
            seen.add(n)
            n = next_val(n)
        return n == 1
